### PR TITLE
Swift 3.0 Preview 1

### PR DESCRIPTION
**Pull request by norio-nomura**
#166
Summary:
- Drop Swift 2.x support
- Swift 3 API Design Guidelines
- Swift 3.0 Preview 1

API changes:
- `Result`
  - `.Success` to `.success`
  - `.Failure` to `.failure`
- `ResultType` to `ResultProtocol`
  - `recoverWith(_:)` to `recover(with:)`
- `ResultErrorType` to `ErrorProtocol`
- `ErrorTypeConvertible` to `ErrorProtocolConvertible`
  - `errorFromErrorType_:)` to `error(from:)`

Migration supporting codes are added to the end of each source files.

### CODE REVIEW

File: Result/ResultProtocol.swift [hunk 1]
```diff
@@ -1,30 +1,20 @@
...
 
-#if swift(>=3.0)
-	public typealias ResultErrorType = ErrorProtocol
-#else
-	public typealias ResultErrorType = ErrorType
-#endif
-
 /// A type that can represent either failure with an error or success with a result value.
-public protocol ResultType {
+public protocol ResultProtocol {
```
Comment (NachoSoto) [inline Result/ResultProtocol.swift]: This change makes me sad

File: Result/ResultProtocol.swift [hunk 2]
```diff
@@ -237,28 +145,40 @@ public func == <T: ResultType where T.Value: Equatable, T.Error: Equatable> (lef
...
-#if swift(>=3)
-public func ?? <T: ResultType> (left: T, right: @autoclosure () -> T) -> T {
-	return left.recoverWith(right())
+public func ?? <T: ResultProtocol> (left: T, right: @autoclosure () -> T) -> T {
+	return left.recover(with: right())
+}
+
+// MARK: - migration support
+@available(*, unavailable, renamed: "ResultProtocol")
+public typealias ResultType = ResultProtocol
```
Comment (NachoSoto) [inline Result/ResultProtocol.swift]: Awesome

**Comment by NachoSoto** [comment]
:shipit:

File: Result/ResultProtocol.swift (hunk 1 above)
Comment (robrix) [inline Result/ResultProtocol.swift]: For why?

Comment (NachoSoto) [inline Result/ResultProtocol.swift]: I think `Type` was a much better suffix, the fact that it's a protocols is already encoded in the type.

Comment (NachoSoto) [inline Result/ResultProtocol.swift]: (Not critiquing this change though, I know it's aligned with the new stdlib conventions.)

Comment (robrix) [inline Result/ResultProtocol.swift]: > I think `Type` was a much better suffix, the fact that it's a protocols is already encoded in the type.

To be fair, the fact that it’s a type is too 😁

> (Not critiquing this change though, I know it's aligned with the new stdlib conventions.)

Understood 👍 I was mostly curious cos I saw the opposite reaction from some folks when they moved to saying `FooType` in the first place.

**Comment by norio-nomura** [comment]
Sorry, I didn't know that migration states are saved on each targets of Xcode project.
I updated `project.pbxproj`.

File: Result.podspec [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Pod::Spec.new do |s|
   s.name         = 'Result'
-  s.version      = '2.1.1'
+  s.version      = '3.0.0-alpha.30p1'
```
Comment (mdiep) [inline Result.podspec]: This should probably be `3.0.0-alpha.1`?

File: Result.podspec [hunk 4]
```diff
@@ -9,7 +9,6 @@ Pod::Spec.new do |s|
   s.source       = { :git => 'https://github.com/antitypical/Result.git', :tag => s.version }
   s.source_files  = 'Result/*.swift'
   s.requires_arc = true
-  s.pod_target_xcconfig = { 'SWIFT_VERSION' => '2.3' }
```
Comment (mdiep) [inline Result.podspec]: This should probably explicitly specify 3.0 (or whatever the correct value is) for the `SWIFT_VERSION`.

File: Result/Info.plist [hunk 5]
```diff
@@ -15,7 +15,7 @@
 	<key>CFBundlePackageType</key>
 	<string>FMWK</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.1.1</string>
+	<string>3.0.0-alpha.30p1</string>
```
Comment (mdiep) [inline Result/Info.plist]: As above, I think `3.0.0-alpha.1` is probably more appropriate.

File: Result/Result.swift [hunk 6]
```diff
@@ -262,6 +173,26 @@ extension NSError: ErrorTypeConvertible {
...
+		fatalError()
+	}
+}
+
+extension NSError {
+	@available(*, unavailable, renamed: "error(from:)")
+	public static func errorFromErrorType(_ error: ErrorProtocol) -> Self {
+		fatalError()
+	}
+}
```
Comment (mdiep) [inline Result/Result.swift]: 💖

**Comment by mdiep** [comment]
I think we should choose a simpler version number, as noted above.

And the tests didn't pass, but I think that may be because Travis has had problems today.

Once those are addressed, I'm 👍 to ship!

**Comment by norio-nomura** [comment]
Thanks all for reviews!
I updated.

**Comment by mdiep** [comment]
💖

**Comment by norio-nomura** [comment]
Thanks! 🙏

**Pull request merged**